session.go

```diff
@@ -31,182 +32,131 @@ const (
-	process, err := s.execer.Start(ctx, Command{
-		Command:    "screen",
-		Args:       append([]string{"-S", id, "-Dmc", s.configFile, s.command.Command}, s.command.Args...),
-		UID:        s.command.UID,
-		GID:        s.command.GID,
-		Env:        append(s.command.Env, "SCREENDIR="+s.socketsDir),
-		WorkingDir: s.command.WorkingDir,
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	// Screen allows targeting sessions via either the session name or
-	// <pid>.<session-name>.  Using the latter form allows us to differentiate
-	// between sessions with the same name.  For example if a session closes due
-	// to a timeout and a client reconnects around the same time there can be two
-	// sessions with the same name while the old session is cleaning up.  This is
-	// only a problem while attaching and not creating since the creation command
-	// used always creates a new session.
-	s.id = fmt.Sprintf("%d.%s", process.Pid(), id)
-	return process, nil
-}
-
-// waitReady waits for the session to be ready by running a command against the
-// session until it works since sometimes if you attach too quickly after
-// spawning screen it will say the session does not exist.  If the provided
-// context finishes the wait is aborted and the context's error is returned.
-func (s *Session) waitReady(ctx context.Context) error {
-	check := func() (bool, error) {
-		// The `version` command seems to be the only command without a side effect
-		// so use it to check whether the session is up.
+// sendCommand runs a screen command against a session.  If the command is quit
+// then the command failing due to the session being missing is considered a
+// success state.  The command will be retried until successful, the timeout is
+// reached, or the context ends (in which case the context error is returned).
+func (s *Session) sendCommand(ctx context.Context, command string) error {
+	ctx, cancel := context.WithTimeout(ctx, s.attachTimeout)
+	defer cancel()
+	run := func() (bool, error) {
 		process, err := s.execer.Start(ctx, Command{
 			Command: "screen",
-			Args:    []string{"-S", s.id, "-X", "version"},
+			Args:    []string{"-S", s.id, "-X", command},
 			UID:     s.command.UID,
 			GID:     s.command.GID,
 			Env:     append(s.command.Env, "SCREENDIR="+s.socketsDir),
 		})
 		if err != nil {
 			return true, err
 		}
+		stdout := captureStdout(process)
 		err = process.Wait()
 		// Try the context error in case it canceled while we waited.
 		if ctx.Err() != nil {
 			return true, ctx.Err()
 		}
-		// Session is ready once we executed without an error.
-		// TODO: Should we specifically check for "no screen to be attached
-		// matching"?  That might be the only error we actually want to retry and
-		// otherwise we return immediately.  But this text may not be stable between
-		// screen versions or there could be additional errors that can be retried.
+		details := <-stdout
+		if strings.Contains(details, "No screen session found") && command == "quit" {
```

Comment:
special casing based on command smells funny.

what about passing a slice of error strings to consider "success" to sendCommand?

Reply:
I like that!